# Introducción a la Ciencia de Datos: Tarea 1

## Análisisi exploratorio sobre la obra de Shakespeare

Este notebook contiene el código utilizado para analizar la base de datos. 

## Cargar bibliotecas (dependencias)

Recuerde instalar los requerimientos (requirements.txt) en el mismo entorno donde está ejecutando este notebook (ver README).

In [38]:
from time import time
from pathlib import Path

import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine


## Conexión a la base de datos y lectura de datos

La creación del directorio /data es opcional, en esa carpeta se guardarán los archivos .csv con las tablas.

In [24]:
# Creamos el directorio Tarea_1/data/shakespeare
data_dir = Path("data") / "shakespeare"
data_dir.mkdir(parents=True, exist_ok=True)


def load_table(table_name, engine):
    """
    Leer la tabla con SQL y guardarla como CSV,
    o cargarla desde el CSV si ya existe
    """
    path_table = data_dir / f"{table_name}.csv"
    if not path_table.exists():
        print(f"Consultando tabla con SQL: {table_name}")
        t0 = time()
        df_table = pd.read_sql(f"SELECT * FROM {table_name}", engine)
        t1 = time()
        print(f"Tiempo: {t1 - t0:.1f} segundos")

        print(f"Guardando: {path_table}\n")
        df_table.to_csv(path_table)
    else:
        print(f"Cargando tabla desde CSV: {path_table}")
        df_table = pd.read_csv(path_table, index_col=[0])
    return df_table


#conectamos con la base de datos

print("Conectando a la base...")
conn_str = "mysql+pymysql://guest:relational@relational.fit.cvut.cz:3306/Shakespeare"
engine = create_engine(conn_str)

# Cargamos las tablas

df_works = load_table("works", engine)

df_paragraphs = load_table("paragraphs", engine)

df_chapters = load_table("chapters", engine)

df_characters = load_table("characters", engine)

Conectando a la base...
Cargando tabla desde CSV: data/shakespeare/works.csv
Cargando tabla desde CSV: data/shakespeare/paragraphs.csv
Cargando tabla desde CSV: data/shakespeare/chapters.csv
Cargando tabla desde CSV: data/shakespeare/characters.csv


## Exploración de Datos

In [5]:
# Veamos las obras incluídas:
df_works


id                      Title                                  LongTitle  \
0    1              Twelfth Night            Twelfth Night, Or What You Will   
1    2  All's Well That Ends Well                  All's Well That Ends Well   
2    3       Antony and Cleopatra                       Antony and Cleopatra   
3    4             As You Like It                             As You Like It   
4    5           Comedy of Errors                       The Comedy of Errors   
5    6                 Coriolanus                                 Coriolanus   
6    7                  Cymbeline                 Cymbeline, King of Britain   
7    8                     Hamlet  Tragedy of Hamlet, Prince of Denmark, The   
8    9           Henry IV, Part I                History of Henry IV, Part I   
9   10          Henry IV, Part II               History of Henry IV, Part II   
10  11                    Henry V                         History of Henry V   
11  12           Henry VI, Part I                History of Henry VI, Part I   
12  13          Henry VI, Part II               History of Henry VI, Part II   
13  14         Henry VI, Part III              History of Henry VI, Part III   
14  15                 Henry VIII                      History of Henry VIII   
15  16              Julius Caesar               The Tragedy of Julius Caesar   
16  17                  King John                       History of King John   
17  18                  King Lear                   The Tragedy of King Lear   
18  19          Lover's Complaint                        A Lover's Complaint   
19  20       Love's Labour's Lost                       Love's Labour's Lost   
20  21                    Macbeth                     The Tragedy of Macbeth   
21  22        Measure for Measure                        Measure for Measure   
22  23         Merchant of Venice                    Merchant of Venice, The   
23  24     Merry Wives of Windsor                Merry Wives of Windsor, The   
24  25    Midsummer Night's Dream                  A Midsummer Night's Dream   
25  26     Much Ado about Nothing                     Much Ado about Nothing   
26  27                    Othello     The Tragedy of Othello, Moor of Venice   
27  28         Passionate Pilgrim                     The Passionate Pilgrim   
28  29                   Pericles                   Pericles, Prince of Tyre   
29  30     Phoenix and the Turtle                 The Phoenix and the Turtle   
30  31            Rape of Lucrece                        The Rape of Lucrece   
31  32                 Richard II                      History of Richard II   
32  33                Richard III                     History of Richard III   
33  34           Romeo and Juliet            The Tragedy of Romeo and Juliet   
34  35                    Sonnets                                    Sonnets   
35  36        Taming of the Shrew                    The Taming of the Shrew   
36  37                    Tempest                                The Tempest   
37  38            Timon of Athens             The Tragedy of Timon of Athens   
38  39           Titus Andronicus                           Titus Andronicus   
39  40       Troilus and Cressida                       Troilus and Cressida   
40  41    Two Gentlemen of Verona                    Two Gentlemen of Verona   
41  42           Venus and Adonis                           Venus and Adonis   
42  43          The Winter's Tale                          The Winter's Tale   

    Date GenreType  
0   1599    Comedy  
1   1602    Comedy  
2   1606   Tragedy  
3   1599    Comedy  
4   1589    Comedy  
5   1607   Tragedy  
6   1609   History  
7   1600   Tragedy  
8   1597   History  
9   1597   History  
10  1598   History  
11  1591   History  
12  1590   History  
13  1590   History  
14  1612   History  
15  1599   Tragedy  
16  1596   History  
17  1605   Tragedy  
18  1609      Poem  
19  1594    Comedy  
20  1605   Tragedy  
21  1604    Comedy  


In [6]:
# parrafos
df_paragraphs

id  ParagraphNum                                          PlainText  \
0  630863             3  [Enter DUKE ORSINO, CURIO, and other Lords; Mu...   
1  630864             4  If music be the food of love, play on;\nGive m...   
2  630865            19                         Will you go hunt, my lord?   
3  630866            20                                       What, Curio?   
4  630867            21                                          The hart.   
5  630868            22  Why, so I do, the noblest that I have:\nO, whe...   
6  630869            30  So please my lord, I might not be admitted;\nB...   
7  630870            39  O, she that hath a heart of that fine frame\nT...   
8  630871            48                                           [Exeunt]   
9  630872            50              [Enter VIOLA, a Captain, and Sailors]   

   character_id  chapter_id  
0          1261       18704  
1           840       18704  
2           297       18704  
3           840       18704  
4           297       18704  
5           840       18704  
6          1221       18704  
7           840       18704  
8          1261       18704  
9          1261       18705

In [7]:
# capitulos
df_chapters

id  Act  Scene                          Description  work_id
0    18704    1      1                DUKE ORSINO’s palace.        1
1    18705    1      2                       The sea-coast.        1
2    18706    1      3                      OLIVIA’S house.        1
3    18707    1      4                DUKE ORSINO’s palace.        1
4    18708    1      5                      OLIVIA’S house.        1
..     ...  ...    ...                                  ...      ...
940  19644    4      3  A road near the Shepherd’s cottage.       43
941  19645    4      4              The Shepherd’s cottage.       43
942  19646    5      1           A room in LEONTES’ palace.       43
943  19647    5      2              Before LEONTES’ palace.       43
944  19648    5      3         A chapel in PAULINA’S house.       43

[945 rows x 5 columns]

In [25]:
# personajes
df_characters

id           CharName             Abbrev         Description
0        1   First Apparition   First Apparition                 NaN
1        2      First Citizen      First Citizen                 NaN
2        3  First Conspirator  First Conspirator                 NaN
3        4    First Gentleman    First Gentleman                 NaN
4        5         First Goth         First Goth                 NaN
...    ...                ...                ...                 ...
1261  1262       Duke of York               YORK  cousin to the king
1262  1263     Young Clifford     YOUNG CLIFFORD                 NaN
1263  1264   Young Coriolanus   Young CORIOLANUS   son to Coriolanus
1264  1265       Young Lucius       Young LUCIUS       son to Lucius
1265  1266       Young Siward       YOUNG SIWARD        Siward's son

[1266 rows x 4 columns]

### Tablas
 - `works`: información detallada de las obras de Shakespeare. Tiene 43 filas y 5 columnas:
    - `id` identificador de la obra
    - `Title` título de la obra (versión corta)
    - `LongTitle` titulo completo de la obra
    - `Date` año de publicación de la obra
    - `GenreType` género literario de la obra

 - `chapters`: identifica los diferentes actos y escenas dentro de cada obra. Tiene 945 filas y 5 columnas:
    - `id` identificador del capítulo
    - `Act`número del Acto
    - `Scene` número de la escena
    - `Description` breve descrićión o título de la escena
    - `work_id` relaciona el `id` de la obra (relación con tabla `works`)
  
- `characters`: identifica los diferentes personajes que intervienen en las obras. Tiene 1266 filas y 4 columnas:
    - `id` identificador del personaje
    - `CharName` Nombre del o los personajes
    - `Abbrev` abreviación del nombre
    - `Description` información que describe al personaje


- `paragraphs`: Contiene el texto principal de las obras. Tiene 35465 y 5 columnas:
    - `id` identificador del párrafo
    - `ParagraphNum` número de párrafo dentro de la obra
    - `PlainText` texto real del párrafo
    - `character_id`indica qué personaje está hablando en el párrafo (si corresponde) (relación con tabla `character`)
    - `chapter_id` indica en qué  capítulo (acto y escena) se encuentra el párrafo (relación con tabla `chapters`)

## Calidad de datos

- Detección de valores null o na
- Problema de idioma (palabras con ´)
- Detección de signos de puntuación

In [39]:
# Contar los valores nulos en paragraph
print(df_paragraphs.isna().sum())

# Contar los valores nulos en character
print(df_characters.isna().sum())

# Contar los valores nulos en chapter
print(df_chapters.isna().sum())

# Contar los valores nulos en works
print(df_works.isna().sum())


id              0
ParagraphNum    0
PlainText       0
character_id    0
chapter_id      0
dtype: int64
id               0
CharName         0
Abbrev           5
Description    646
dtype: int64
id             0
Act            0
Scene          0
Description    0
work_id        0
dtype: int64
id           0
Title        0
LongTitle    0
Date         0
GenreType    0
dtype: int64


In [34]:
import string

def puntuacion(df, column_name):
    # Obtener el texto de la columna
    texto = df[column_name].str.cat(sep=" ") #str.cat concatena todo el texto de la columna
    #separando por un espacio

    # Encontrar los signos de puntuación
    puntuacion = []
    for char in texto:
        if char in string.punctuation and char not in puntuacion: 
            #string.punctuaction  es una cadena predefinida en la biblioteca estándar de Python
            # que contiene todos los caracteres de puntuación
            puntuacion.append(char)
    
    return puntuacion

# Obtener una lista de signos de puntuación en la columna "PlainText" de df_paragraph
listaParag = puntuacion(df_paragraphs, "PlainText")
print(listaParag)

['[', ',', ';', ']', '.', '!', ':', "'", '?', '-', '&', '(', ')', '"']


In [35]:
#Obtener una lista de signos de puntuación en la columna "CharName" de df_characters
listaNames = puntuacion(df_characters, "CharName")
print(listaNames)

['-', ',', '(', ')', "'", '/']


In [36]:
#Obtener una lista de signos de puntuación en la columna "Description" de df_chapters
listaChap = puntuacion(df_chapters, "Description")
print(listaChap)

['.', '-', ',', ':', '&', ';', '[', ']']


## Limpieza de Texto y Conteo de Palabras

In [37]:
def clean_text(df, column_name): #defino la función clean_text con los parametros
    #df: el data frame que quiero y column_name, la columna que quiero
    # Convertir todo a minúsculas
    result = df[column_name].str.lower()

    # Quitar signos de puntuación y cambiarlos por espacios (" ")
    for punc in ["[", "\n", "," , ";", "]", ".", ":", "'", "?", "-", "&", "(", ")", '"',
                 "-", "/", "[", "’"]: #acá agregamos los signos que detectamos antes
        result = result.str.replace(punc, " ")
    return result

In [40]:
# Creamos una nueva columna CleanText a partir de PlainText en paragraphs
df_paragraphs["CleanText"] = clean_text(df_paragraphs, "PlainText")

# Veamos la diferencia
df_paragraphs[["PlainText", "CleanText"]]

/tmp/ipykernel_28049/2901183826.py:9: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  result = result.str.replace(punc, " ")


PlainText  \
0      [Enter DUKE ORSINO, CURIO, and other Lords; Mu...   
1      If music be the food of love, play on;\nGive m...   
2                             Will you go hunt, my lord?   
3                                           What, Curio?   
4                                              The hart.   
...                                                  ...   
35460  That she is living,\nWere it but told you, sho...   
35461  You gods, look down\nAnd from your sacred vial...   
35462  There's time enough for that;\nLest they desir...   
35463  O, peace, Paulina!\nThou shouldst a husband ta...   
35464                                           [Exeunt]   

                                               CleanText  
0       enter duke orsino  curio  and other lords  mu...  
1      if music be the food of love  play on  give me...  
2                             will you go hunt  my lord   
3                                           what  curio   
4                                              the hart   
...                                                  ...  
35460  that she is living  were it but told you  shou...  
35461  you gods  look down and from your sacred vials...  
35462  there s time enough for that  lest they desire...  
35463  o  peace  paulina! thou shouldst a husband tak...  
35464                                            exeunt   

[35465 rows x 2 columns]

In [41]:
# Creamos una nueva columna CleanText a partir de CharName en df_characters
df_characters["CleanText"] = clean_text(df_characters, "CharName")

# Creamos una nueva columna CleanText a partir de Description en df_chapters
df_chapters["CleanText"] = clean_text(df_chapters, "Description")

# Creamos una nueva columna CleanText a partir de Title en df_works
df_works["CleanText"] = clean_text(df_works, "Title")

/tmp/ipykernel_28049/2901183826.py:9: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  result = result.str.replace(punc, " ")
/tmp/ipykernel_28049/2901183826.py:9: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  result = result.str.replace(punc, " ")
/tmp/ipykernel_28049/2901183826.py:9: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  result = result.str.replace(punc, " ")


### Conteo de palabras

In [42]:
# Convierte párrafos en listas "palabra1 palabra2 palabra3" -> ["palabra1", "palabra2", "palabra3"]
df_paragraphs["WordList"] = df_paragraphs["CleanText"].str.split()

# Veamos la nueva columna creada
# Notar que a la derecha tenemos una lista: [palabra1, palabra2, palabra3]
df_paragraphs[["CleanText", "WordList"]]

CleanText  \
0       enter duke orsino  curio  and other lords  mu...   
1      if music be the food of love  play on  give me...   
2                             will you go hunt  my lord    
3                                           what  curio    
4                                              the hart    
...                                                  ...   
35460  that she is living  were it but told you  shou...   
35461  you gods  look down and from your sacred vials...   
35462  there s time enough for that  lest they desire...   
35463  o  peace  paulina! thou shouldst a husband tak...   
35464                                            exeunt    

                                                WordList  
0      [enter, duke, orsino, curio, and, other, lords...  
1      [if, music, be, the, food, of, love, play, on,...  
2                        [will, you, go, hunt, my, lord]  
3                                          [what, curio]  
4                                            [the, hart]  
...                                                  ...  
35460  [that, she, is, living, were, it, but, told, y...  
35461  [you, gods, look, down, and, from, your, sacre...  
35462  [there, s, time, enough, for, that, lest, they...  
35463  [o, peace, paulina!, thou, shouldst, a, husban...  
35464                                           [exeunt]  

[35465 rows x 2 columns]

In [43]:
# Nuevo dataframe: cada fila ya no es un párrafo, sino una sóla palabra
df_words = df_paragraphs.explode("WordList")

# Quitamos estas columnas redundantes
df_words.drop(columns=["CleanText", "PlainText"], inplace=True)

# Renombramos la columna WordList -> word
df_words.rename(columns={"WordList": "word"}, inplace=True)

# Verificar que el número de filas es mucho mayor
df_words

id  ParagraphNum  character_id  chapter_id     word
0      630863             3          1261       18704    enter
0      630863             3          1261       18704     duke
0      630863             3          1261       18704   orsino
0      630863             3          1261       18704    curio
0      630863             3          1261       18704      and
...       ...           ...           ...         ...      ...
35463  666326          3483           667       19648        d
35463  666326          3483           667       19648  hastily
35463  666326          3483           667       19648     lead
35463  666326          3483           667       19648     away
35464  666327          3504          1261       19648   exeunt

[909372 rows x 5 columns]

In [44]:
# conteo inicial de palabras

df_count_words = df_words['word'].value_counts()
df_count_words#.head(10)

the          28933
and          27312
i            22994
to           20812
of           17174
             ...  
writs            1
eternized        1
gaunt!           1
poltroons        1
directing        1
Name: word, Length: 26359, dtype: int64

## Cantidad de párrafos por personaje. 
¿Cuál es el personaje con más párrafos?

In [45]:
# tabla paragraph, agrupo por personaje y por chapter_id

df_agrupada = df_paragraphs.groupby(['character_id','chapter_id'])['id'].count()
result = df_agrupada.rename('paragraphs').sort_values(ascending=False).reset_index()
print(result)

      character_id  chapter_id  paragraphs
0              894       19308         265
1              894       19633         197
2              573       18901          78
3             1198       19568          74
4              945       19346          73
...            ...         ...         ...
4871           894       19440           1
4872           894       19439           1
4873           894       19438           1
4874           894       19437           1
4875           742       19304           1

[4876 rows x 3 columns]


In [46]:
#quiero saber el nombre del personaje, hago merge con df_characters
df_par_name = pd.merge(result, df_characters[["id", "CharName", "CleanText"]], left_on="character_id", right_on="id").sort_values('paragraphs', ascending=False)
df_par_name

character_id  chapter_id  paragraphs    id     CharName    CleanText
0              894       19308         265   894         Poet         poet
1              894       19633         197   894         Poet         poet
182            573       18901          78   573      Henry V      henry v
208           1198       19568          74  1198        Timon        timon
216            945       19346          73   945  Richard III  richard iii
...            ...         ...         ...   ...          ...          ...
3769            86       19156           1    86          All          all
3770            86       19131           1    86          All          all
3771            86       19129           1    86          All          all
3789           398       19216           1   398       Fenton       fenton
4875           815       19540           1   815     Nicholas     nicholas

[4876 rows x 6 columns]

In [47]:
# me interesa saber si están en la misma obra, hago merge con chapter
df_merg_chap = pd.merge(df_chapters, df_par_name[["id", "chapter_id", "character_id", "paragraphs", "CharName", "CleanText"]], left_on="id", right_on="chapter_id").sort_values('paragraphs', ascending=False)
df_merg_chap

id_x  Act  Scene                         Description  work_id  \
3558  19308    1      2                 The Rape of Lucrece       31   
4786  19633    1      1                   Venus and Adonis.       42   
1131  18901    2      4  The Boar’s-Head Tavern, Eastcheap.        9   
4411  19568    4      3  Woods and cave, near the seashore.       38   
3790  19346    4      4                  Before the palace.       33   
...     ...  ...    ...                                 ...      ...   
4110  19506    1    128                               ---\n       35   
4111  19507    1    129                               ---\n       35   
4112  19508    1    130                               ---\n       35   
4113  19509    1    131                               ---\n       35   
4875  19648    5      3        A chapel in PAULINA’S house.       43   

                             CleanText_x  id_y  chapter_id  character_id  \
3558                 the rape of lucrece   894       19308           894   
4786                   venus and adonis    894       19633           894   
1131  the boar s head tavern  eastcheap    573       18901           573   
4411  woods and cave  near the seashore   1198       19568          1198   
3790                  before the palace    945       19346           945   
...                                  ...   ...         ...           ...   
4110                                       894       19506           894   
4111                                       894       19507           894   
4112                                       894       19508           894   
4113                                       894       19509           894   
4875        a chapel in paulina s house   1261       19648          1261   

      paragraphs            CharName         CleanText_y  
3558         265                Poet                poet  
4786         197                Poet                poet  
1131          78             Henry V             henry v  
4411          74               Timon               timon  
3790          73         Richard III         richard iii  
...          ...                 ...                 ...  
4110           1                Poet                poet  
4111           1                Poet                poet  
4112           1                Poet                poet  
4113           1                Poet                poet  
4875           1  (stage directions)   stage directions   

[4876 rows x 12 columns]

In [48]:
# me interesa saber si están en la misma obra, hago merge con work para traer el titulo
df_merg_work = pd.merge(df_merg_chap, df_works[["id", "Title"]], left_on="work_id", right_on="id").sort_values('paragraphs', ascending=False)

#selecciono las columnas importantes
df_merg_work = df_merg_work.loc[:, ["work_id", "chapter_id", "character_id", "paragraphs", "CharName", "Title"]]

df_merg_work.head(10) #10 personajes con mayor cantidad de párrafos, discriminados por obra

work_id  chapter_id  character_id  paragraphs     CharName  \
0         31       19308           894         265         Poet   
3         42       19633           894         197         Poet   
5          9       18901           573          78      Henry V   
114       38       19568          1198          74        Timon   
231       33       19346           945          73  Richard III   
506       16       19071           201          69       Brutus   
400       27       19257           844          69      Othello   
639       40       19589           292          68     Cressida   
401       27       19251           844          67      Othello   
795       20       19128           170          67        Biron   

                    Title  
0         Rape of Lucrece  
3        Venus and Adonis  
5        Henry IV, Part I  
114       Timon of Athens  
231           Richard III  
506         Julius Caesar  
400               Othello  
639  Troilus and Cressida  
401               Othello  
795  Love's Labour's Lost